source/kernel/src/bbq.rs

```diff
@@ -84,14 +106,167 @@ impl Drop for BBQStorage {
+        self.grant.deref()
     }
+}
+
+impl DerefMut for GrantW {
+    #[inline(always)]
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.grant.deref_mut()
+    }
+}
+
+impl GrantW {
+    pub fn commit(self, used: usize) {
+        self.grant.commit(used);
+        // If we freed up any space, notify the waker on the reader side
+        if used != 0 {
+            match self.side {
+                Side::ASide => &self.storage.a_wait,
+                Side::BSide => &self.storage.b_wait,
+            }.commit_waitcell.notify();
+        }
+    }
+}
+
+pub struct GrantR {
+    grant: InnerGrantR<'static>,
+    storage: HeapArc<BBQStorage>,
+    side: Side,
+}
+
+impl Deref for GrantR {
+    type Target = [u8];
 
     #[inline(always)]
-    pub fn consumer(&self) -> &Consumer<'static> {
-        &self.consumer
+    fn deref(&self) -> &Self::Target {
+        self.grant.deref()
+    }
+}
+
+impl DerefMut for GrantR {
+    #[inline(always)]
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.grant.deref_mut()
+    }
+}
+
+impl GrantR {
+    pub fn release(self, used: usize) {
+        self.grant.release(used);
+        // If we freed up any space, notify the waker on the reader side
+        if used != 0 {
+            match self.side {
+                Side::ASide => &self.storage.a_wait,
+                Side::BSide => &self.storage.b_wait,
+            }.release_waitcell.notify();
+        }
+    }
+}
+
+impl BBQBidiHandle {
+    // async fn send_grant(buf_len: usize) -> GrantW
+    // async fn read_grant() -> GrantR
+    pub async fn send_grant_max(&self, max: usize) -> GrantW {
+        loop {
+            match self.producer.grant_max_remaining(max) {
+                Ok(wgr) => {
+                    trace!(
+                        size = wgr.len(),
+                        max = max,
+                        "Got bbqueue max write grant",
+                    );
+                    return GrantW {
+                        grant: wgr,
+                        side: self.side,
+                        storage: self.storage.clone(),
+                    }
+                },
+                Err(_) => {
+                    trace!("awaiting bbqueue max write grant");
+                    // Uh oh! Couldn't get a send grant. We need to wait for the OTHER reader
+                    // to release some bytes first.
+                    match self.side {
+                        Side::ASide => &self.storage.b_wait,
+                        Side::BSide => &self.storage.a_wait,
+                    }.release_waitcell.wait().await.unwrap();
+
+                    trace!("awoke for bbqueue max write grant");
+                },
+            }
+        }
+    }
+
+    pub async fn send_grant_exact(&self, size: usize) -> GrantW {
+        loop {
+            match self.producer.grant_exact(size) {
+                Ok(wgr) => {
+                    trace!(
+                        size = size,
```

Comment:
tiny nit: this could just be
```suggestion
                        size,
```